Makefile

```
@@ -0,0 +1,138 @@
+# Directories.
+APIS_DIR := pkg/apis
+CLIENTSET_DIR := pkg/clientset
+CRDV1_DIR := config/crd/v1
+CRDV1BETA1_DIR := config/crd/v1beta1
+SCRIPTS_DIR := hack
+TOOLS_DIR := $(SCRIPTS_DIR)/tools
+TOOLS_BIN_DIR := $(abspath $(TOOLS_DIR)/bin)
+
+# Binaries.
+# Need to use abspath so we can invoke these from subdirectories
+CLIENT_GEN := $(abspath $(TOOLS_BIN_DIR)/client-gen)
+CONTROLLER_GEN := $(abspath $(TOOLS_BIN_DIR)/controller-gen)
+DEEPCOPY_GEN := $(abspath $(TOOLS_BIN_DIR)/deepcopy-gen)
+GOIMPORTS := $(abspath $(TOOLS_BIN_DIR)/goimports)
+GOLANGCI_LINT := $(abspath $(TOOLS_BIN_DIR)/golangci-lint)
+KUSTOMIZE := $(abspath $(TOOLS_BIN_DIR)/kustomize)
+ESC := $(abspath $(TOOLS_BIN_DIR)/esc)
+
+BUILD_COLOR = \033[0;34m
+GEN_COLOR = \033[0;32m
+
+DEEPCOPY_BASE = zz_generated.deepcopy
+MODULE = github.com/giantswarm/apiextensions
+BOILERPLATE = $(SCRIPTS_DIR)/boilerplate.go.txt
+PATCH_FILE = $(SCRIPTS_DIR)/generated.patch
+
+INPUT_DIRS := $(shell find ./$(APIS_DIR) -maxdepth 2 -mindepth 2 | paste -s -d, -)
+GROUPS := $(shell find $(APIS_DIR) -maxdepth 2 -mindepth 2  | sed 's|pkg/apis/||' | paste -s -d, -)
+DEEPCOPY_FILES := $(shell find $(APIS_DIR) -name $(DEEPCOPY_BASE).go)
+
+all: generate
+
+$(CLIENT_GEN): $(TOOLS_DIR)/client-gen/go.mod
+	@echo "$(BUILD_COLOR)Building client-gen"
+	@cd $(TOOLS_DIR)/client-gen; go build -tags=tools -o $(CLIENT_GEN) k8s.io/code-generator/cmd/client-gen
+
+$(CONTROLLER_GEN): $(TOOLS_DIR)/controller-gen/go.mod
+	@echo "$(BUILD_COLOR)Building controller-gen"
+	@cd $(TOOLS_DIR)/controller-gen; go build -tags=tools -o $(CONTROLLER_GEN) sigs.k8s.io/controller-tools/cmd/controller-gen
+
+$(DEEPCOPY_GEN): $(TOOLS_DIR)/deepcopy-gen/go.mod
+	@echo "$(BUILD_COLOR)Building deepcopy-gen"
+	@cd $(TOOLS_DIR)/deepcopy-gen; go build -tags=tools -o $(DEEPCOPY_GEN) k8s.io/code-generator/cmd/deepcopy-gen
+
+$(GOIMPORTS): $(TOOLS_DIR)/goimports/go.mod
+	@echo "$(BUILD_COLOR)Building goimports"
+	@cd $(TOOLS_DIR)/goimports; go build -tags=tools -o $(GOIMPORTS) golang.org/x/tools/cmd/goimports
+
+$(GOLANGCI_LINT): $(TOOLS_DIR)/golangci-lint/go.mod
+	@echo "$(BUILD_COLOR)Building golangci-lint"
+	@cd $(TOOLS_DIR)/golangci-lint; go build -tags=tools -o $(GOLANGCI_LINT) github.com/golangci/golangci-lint/cmd/golangci-lint
+
+$(KUSTOMIZE): $(TOOLS_DIR)/kustomize/go.mod
+	@echo "$(BUILD_COLOR)Building kustomize"
+	@cd $(TOOLS_DIR)/kustomize; go build -tags=tools -o $(KUSTOMIZE) sigs.k8s.io/kustomize/kustomize/v3
+
+$(ESC): $(TOOLS_DIR)/esc/go.mod
+	@echo "$(BUILD_COLOR)Building esc"
+	@cd $(TOOLS_DIR)/esc; go build -tags=tools -o $(ESC) github.com/mjibson/esc
+
+.PHONY: generate
+generate:
+	@$(MAKE) generate-clientset
+	@$(MAKE) generate-deepcopy
+	@$(MAKE) generate-manifests
+	@$(MAKE) generate-static
+	@$(MAKE) imports
+	@$(MAKE) patch
+
+.PHONY: verify
+verify:
+	@$(MAKE) clean-generated
+	@$(MAKE) generate
+	@$(MAKE) lint
+	@git diff --exit-code
```

It's useful to see the command in this case IMO.

```suggestion
	git diff --exit-code
```